File: apps/mobile/ios/SpeakeasyBridges/Signal/SpeakeasySignalStore.swift
//
//  SpeakeasySignalStore.swift
//  Speakeasy
//
//  Singleton holder for the active SignalProtocol store. Mirrors
//  apps/mobile/android/.../signal/SpeakeasySignalStore.kt
//
//  Phase 5c: SQLCipher-backed by default. On first init the store is
//  reconstructed from disk (`loadFromDb`); on a fresh install the
//  caller passes a freshly generated identity to `initialize` which
//  writes it to the encrypted DB.
//

import Foundation
import LibSignalClient

enum SpeakeasySignalStore {
    private static let lock = NSLock()
    private static var instance: SqlCipherSignalProtocolStore?

    /// Restore from disk if a previous identity exists. Returns false ONLY
    /// when the store is genuinely absent — no identity row yet (fresh
    /// install), or the DB can't open because Vouchflow.cachedDeviceToken
    /// is nil (not enrolled). Any OTHER failure THROWS.
    ///
    /// The distinction is load-bearing: `generateIdentityKey` mints a
    /// fresh identity on `false`. The previous version returned false for
    /// unexpected DB failures too (transient keychain unavailability,
    /// SQLCipher open errors during an update relaunch), which silently
    /// REPLACED the user's identity while they kept their handle — every
    /// peer then saw "[identity changed — verify with peer]" and calls
    /// died on UntrustedIdentityException, with no reinstall having
    /// happened. Android's initializeFromDb has always propagated
    /// unexpected errors; this brings iOS to parity.
    static func initializeFromDb() throws -> Bool {
        lock.lock()
        defer { lock.unlock() }
        do {
            let db = try SpeakeasyDb.shared.open()
            guard let store = try SqlCipherSignalProtocolStore.loadFromDb(db) else {
                return false
            }
            instance = store
            return true
        } catch SpeakeasyDbError.notEnrolled {
            return false
        }
        // Unexpected DB failure propagates — callers must NOT treat it
        // as "no identity yet".
    }

    /// Initialise from a freshly generated identity. Persists to the DB.
    static func initialize(identityKeyPair: IdentityKeyPair, registrationId: UInt32) throws {
        lock.lock()
        defer { lock.unlock() }
        let db = try SpeakeasyDb.shared.open()
        let store = SqlCipherSignalProtocolStore(
            db: db,
            identityKeyPair: identityKeyPair,
            registrationId: registrationId
        )
        try store.persistLocalIdentity()
        instance = store
    }

    static var isInitialized: Bool {
        lock.lock(); defer { lock.unlock() }
        return instance != nil
    }

    static func require() throws -> SqlCipherSignalProtocolStore {
        lock.lock(); defer { lock.unlock() }
        guard let i = instance else {
            throw NSError(domain: "SpeakeasySignalStore", code: -1,
                          userInfo: [NSLocalizedDescriptionKey:
                                        "SpeakeasySignalStore not initialized — call generateIdentityKey first"])
        }
        return i
    }

    static func reset() {
        lock.lock(); defer { lock.unlock() }
        instance = nil
    }
}

File: apps/mobile/android/app/src/main/java/xyz/speakeasyapp/app/signal/SpeakeasySignalStore.kt
package xyz.speakeasyapp.app.signal

import android.content.Context
import org.signal.libsignal.protocol.IdentityKeyPair
import org.signal.libsignal.protocol.state.SignalProtocolStore
import org.signal.libsignal.protocol.state.impl.InMemorySignalProtocolStore
import xyz.speakeasyapp.app.db.SpeakeasyDb

/**
 * Singleton holder for the active `SignalProtocolStore`.
 *
 * # Phase 5c: SQLCipher-backed by default
 *
 * On first init the store is reconstructed from disk if a previous
 * identity exists ([SqlCipherSignalProtocolStore.loadFromDb]). On a fresh
 * install, the caller passes a freshly-generated identity to [initialize]
 * which writes it to the encrypted DB.
 *
 * # Test / dev fallback
 *
 * [initializeInMemory] keeps the old behaviour for unit-style harnesses
 * that don't have a `Context` available.
 */
object SpeakeasySignalStore {
  @Volatile private var instance: SignalProtocolStore? = null

  /**
   * Phase 5c: persistent. The DB call ladder is open() → load identity row.
   *
   * Returns `false` in two distinct "not loaded yet" cases:
   *   - DB exists but no identity row yet (fresh install post-enrollment).
   *   - Vouchflow not yet enrolled — DB can't even open without a
   *     `cachedDeviceToken` ([SpeakeasyDb.NotEnrolledException]).
   *
   * Both cases mean "JS layer should mint a fresh identity and call
   * [initialize]." [initialize] requires the deviceToken be present, so
   * the JS layer must run `vouchflow.verify` first.
   */
  fun initializeFromDb(context: Context): Boolean {
    val db =
        try {
          SpeakeasyDb.open(context)
        } catch (_: SpeakeasyDb.NotEnrolledException) {
          return false
        }
    val loaded = SqlCipherSignalProtocolStore.loadFromDb(db) ?: return false
    instance = loaded
    return true
  }

  /** Phase 5c: persistent — used right after generating a fresh identity. */
  fun initialize(context: Context, identityKeyPair: IdentityKeyPair, registrationId: Int) {
    val db = SpeakeasyDb.open(context)
    val store = SqlCipherSignalProtocolStore(db, identityKeyPair, registrationId)
    store.persistLocalIdentity()
    instance = store
  }

  /** Phase 5b legacy: kept for harnesses without an Android Context. */
  fun initializeInMemory(identityKeyPair: IdentityKeyPair, registrationId: Int) {
    instance = InMemorySignalProtocolStore(identityKeyPair, registrationId)
  }

  fun requireInitialized(): SignalProtocolStore =
      instance
          ?: throw IllegalStateException(
              "SpeakeasySignalStore not initialized — call generateIdentityKey first")

  fun isInitialized(): Boolean = instance != null

  /** Drop in-memory handle. The on-disk DB is left intact. */
  fun reset() {
    instance = null
  }
}
